server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java

```diff
@@ -212,6 +221,111 @@ public void testDeleteCurrentWriteIndexForDataStream() {
         );
     }
 
+    public void testDeleteIndexWithSnapshotDeletion() {
+        final boolean deleteSnapshot = randomBoolean();
+        final IndexMetadata indexMetadata = IndexMetadata.builder("test")
+            .settings(
+                Settings.builder()
+                    .put("index.version.created", VersionUtils.randomVersion(random()))
+                    .put(IndexModule.INDEX_STORE_TYPE_SETTING.getKey(), SearchableSnapshotsSettings.SEARCHABLE_SNAPSHOT_STORE_TYPE)
+                    .put(SEARCHABLE_SNAPSHOTS_REPOSITORY_NAME_SETTING_KEY, "repo_name")
+                    .put(SEARCHABLE_SNAPSHOTS_REPOSITORY_UUID_SETTING_KEY, randomBoolean() ? null : "repo_uuid")
+                    .put(SEARCHABLE_SNAPSHOTS_SNAPSHOT_NAME_SETTING_KEY, "snap_name")
+                    .put(SEARCHABLE_SNAPSHOTS_SNAPSHOT_UUID_SETTING_KEY, "snap_uuid")
+                    .put(SearchableSnapshotsSettings.SEARCHABLE_SNAPSHOTS_DELETE_SNAPSHOT_ON_INDEX_DELETION, deleteSnapshot)
+                    .build()
+            )
+            .numberOfShards(1)
+            .numberOfReplicas(1)
+            .build();
+        final ClusterState initialState = ClusterState.builder(ClusterName.DEFAULT)
+            .metadata(
+                Metadata.builder()
+                    .put(indexMetadata, false)
+                    .putCustom(
+                        RepositoriesMetadata.TYPE,
+                        new RepositoriesMetadata(List.of(new RepositoryMetadata("repo_name", "fs", Settings.EMPTY).withUuid("repo_uuid")))
+                    )
+            )
+            .routingTable(RoutingTable.builder().addAsNew(indexMetadata).build())
+            .blocks(ClusterBlocks.builder().addBlocks(indexMetadata))
+            .build();
+
+        final ClusterState updatedState = service.deleteIndices(initialState, Set.of(indexMetadata.getIndex()));
+        assertThat(updatedState.metadata().getIndices().get("test"), nullValue());
+        assertThat(updatedState.blocks().indices().get("test"), nullValue());
+        assertThat(updatedState.routingTable().index("test"), nullValue());
+
+        final SnapshotDeletionsPending updatedPendingDeletions = updatedState.custom(SnapshotDeletionsPending.TYPE);
+        if (deleteSnapshot) {
+            assertThat(updatedPendingDeletions, notNullValue());
+            assertThat(updatedPendingDeletions.isEmpty(), equalTo(false));
```

Comment:
Can we also validate the it has size 1, i.e., only led to one pending deletion?

Reply:
Added in fce436543320a8b2e610280cb8328223e49c3fbd